File: pulsar-core/pulsar-tools/pulsar-browser/src/main/java/com/github/kklisura/cdt/protocol/v2023/commands/Memory.java
package com.github.kklisura.cdt.protocol.v2023.commands;

/*-
 * #%L
 * cdt-java-client
 * %%
 * Copyright (C) 2018 - 2023 Kenan Klisura
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.github.kklisura.cdt.protocol.v2023.support.annotations.Experimental;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.Optional;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.ParamName;
import com.github.kklisura.cdt.protocol.v2023.support.annotations.Returns;
import com.github.kklisura.cdt.protocol.v2023.types.memory.DOMCounters;
import com.github.kklisura.cdt.protocol.v2023.types.memory.PressureLevel;
import com.github.kklisura.cdt.protocol.v2023.types.memory.SamplingProfile;

@Experimental
public interface Memory {

  DOMCounters getDOMCounters();

  void prepareForLeakDetection();

  /** Simulate OomIntervention by purging V8 memory. */
  void forciblyPurgeJavaScriptMemory();

  /**
   * Enable/disable suppressing memory pressure notifications in all processes.
   *
   * @param suppressed If true, memory pressure notifications will be suppressed.
   */
  void setPressureNotificationsSuppressed(@ParamName("suppressed") Boolean suppressed);

  /**
   * Simulate a memory pressure notification in all processes.
   *
   * @param level Memory pressure level of the notification.
   */
  void simulatePressureNotification(@ParamName("level") PressureLevel level);

  /** Start collecting native memory profile. */
  void startSampling();

  /**
   * Start collecting native memory profile.
   *
   * @param samplingInterval Average number of bytes between samples.
   * @param suppressRandomness Do not randomize intervals between samples.
   */
  void startSampling(
      @Optional @ParamName("samplingInterval") Integer samplingInterval,
      @Optional @ParamName("suppressRandomness") Boolean suppressRandomness);

  /** Stop collecting native memory profile. */
  void stopSampling();

  /** Retrieve native memory allocations profile collected since renderer process startup. */
  @Returns("profile")
  SamplingProfile getAllTimeSamplingProfile();

  /** Retrieve native memory allocations profile collected since browser process startup. */
  @Returns("profile")
  SamplingProfile getBrowserSamplingProfile();

  /** Retrieve native memory allocations profile collected since last `startSampling` call. */
  @Returns("profile")
  SamplingProfile getSamplingProfile();
}

File: pulsar-core/pulsar-tools/pulsar-browser/src/main/kotlin/ai/platon/cdt/kt/protocol/commands/Memory.kt
@file:Suppress("unused")
package ai.platon.cdt.kt.protocol.commands

import ai.platon.cdt.kt.protocol.support.annotations.Experimental
import ai.platon.cdt.kt.protocol.support.annotations.Optional
import ai.platon.cdt.kt.protocol.support.annotations.ParamName
import ai.platon.cdt.kt.protocol.support.annotations.Returns
import ai.platon.cdt.kt.protocol.types.memory.DOMCounters
import ai.platon.cdt.kt.protocol.types.memory.PressureLevel
import ai.platon.cdt.kt.protocol.types.memory.SamplingProfile
import kotlin.Boolean
import kotlin.Int

@Experimental
interface Memory {
  suspend fun getDOMCounters(): DOMCounters

  suspend fun prepareForLeakDetection()

  /**
   * Simulate OomIntervention by purging V8 memory.
   */
  suspend fun forciblyPurgeJavaScriptMemory()

  /**
   * Enable/disable suppressing memory pressure notifications in all processes.
   * @param suppressed If true, memory pressure notifications will be suppressed.
   */
  suspend fun setPressureNotificationsSuppressed(@ParamName("suppressed") suppressed: Boolean)

  /**
   * Simulate a memory pressure notification in all processes.
   * @param level Memory pressure level of the notification.
   */
  suspend fun simulatePressureNotification(@ParamName("level") level: PressureLevel)

  /**
   * Start collecting native memory profile.
   * @param samplingInterval Average number of bytes between samples.
   * @param suppressRandomness Do not randomize intervals between samples.
   */
  suspend fun startSampling(@ParamName("samplingInterval") @Optional samplingInterval: Int? = null, @ParamName("suppressRandomness") @Optional suppressRandomness: Boolean? = null)

  suspend fun startSampling() {
    return startSampling(null, null)
  }

  /**
   * Stop collecting native memory profile.
   */
  suspend fun stopSampling()

  /**
   * Retrieve native memory allocations profile
   * collected since renderer process startup.
   */
  @Returns("profile")
  suspend fun getAllTimeSamplingProfile(): SamplingProfile

  /**
   * Retrieve native memory allocations profile
   * collected since browser process startup.
   */
  @Returns("profile")
  suspend fun getBrowserSamplingProfile(): SamplingProfile

  /**
   * Retrieve native memory allocations profile collected since last
   * `startSampling` call.
   */
  @Returns("profile")
  suspend fun getSamplingProfile(): SamplingProfile
}
